### ISSUE TITLE
Compress and shrink the size of the CAS git repository

### ISSUE THREAD

**Issue by mmoayyed** [apereo/cas#1814]
CAS git repo size is reportedly somewhere close to: 609 Mb. 

```
> du -sh .git/
609M    .git/
```

Running `git gc` does help some:

```
> git count-objects -vH
count: 681
size: 417.73 KiB
in-pack: 3109359
packs: 1
size-pack: 593.10 MiB
prune-packable: 0
garbage: 0
size-garbage: 0 bytes
```

Needles to say, this is dysfunctional as clones take "forever" to complete. 

We should try to shrink the git repo size by as much as possible. Here is a good start: 
http://stevelorek.com/how-to-shrink-a-git-repository.html

**Comment by mmoayyed** [comment]
Probably should give this tool a try too:
https://rtyley.github.io/bfg-repo-cleaner/

Though I doubt we have any files that are over 1Mb.

**Comment by mmoayyed** [comment]
- Archive the existing repo somewhere before cleaning and compacting things up. 
- Advertise the task on appropriate channels before starting with this. 
- Explain the process in such a way to the developer community so that they understand history is rewritten, but also emphasize that no work is lost and the state of the project remains in tact as far as features, bug fixes, etc.

**Comment by alexkolson** [comment]
I'll preface my comments by saying I wouldn't be sure if it's worth the effort, but what about splitting each component of CAS into it's own git repository? [Linus himself said the following](http://osdir.com/ml/git/2009-05/msg00051.html):

> git scales really badly if you force it to look at everything as one _huge_ repository

 You could have some sort of aggregation repository that has the components all linked in as submodules for record-keeping purposes. You guys have already done a fantastic job of separating CAS into smaller logical components that are mostly organized in separate top-level subfolders in the repository. What about taking it one step further and making each one of these subfolders its own repository?

The actual [git repo separation process](https://help.github.com/articles/splitting-a-subfolder-out-into-a-new-repository/) is fairly straightforward, but heaven knows what kind of changes might need to then take place in the build/overlay process and of course, as stated before, it'd be a fairly large effort.

**Comment by mmoayyed** [comment]
Yes, I don't disagree. While that may be ideal, personally I am not sure if THAT is worth the effort. @leleuj and I have discussed this briefly in the past. It's a HUGE undertaking. To be precise, we only worry about the initial clones. That's the goal. That is done once and you sort of forget it then. Compartmentalizing the entire project goes way beyond that because as you said, we have to worry about docs, overlays, release management, maintenance, developer education, etc. I'd argue differently if we had a very large developer community and were receiving 4-5 PRs every day. I'd also argue differently if we were presenting an SDK. (If your target audience are pretty much developers and Java ninjas, you generally don't have to worry about all of that. They will just get it) The pain is only felt by a only few in comparison, and so this is a plan for the time being to navigate around that. 

Who knows. Maybe in a few years we will all have moved onto something other than git that scales much better...or maybe the CAS architecture evolves in such a way that the project reorganization might be unnecessary. I'd want to wait this out a while longer and see what the tool can do for us preferably rather than us solving its "problems".
